mogwai/src/view/dom.rs

```diff
@@ -62,407 +67,261 @@ pub(crate) struct ServerNode {
-            .expect(&format!("cannot create element {:?}", name))
-            .unchecked_into();
-        View::wrapping(el)
-    }
 
-    #[cfg(not(target_arch = "wasm32"))]
-    fn element_ns(tag: &str, ns: &str) -> Self {
-        View {
-            element: Rc::new(JsValue::NULL),
-            phantom: std::marker::PhantomData,
-            server_node: ServerNode {
-                name_or_text: NameOrText::Name(Rc::new(RefCell::new(tag.into()))),
-                attributes: vec![("xmlns".into(), Rc::new(RefCell::new(Some(ns.into()))))],
-                styles: vec![],
-            },
-            children: vec![],
-
-            callbacks: HashMap::default(),
-            window_callbacks: HashMap::default(),
-            document_callbacks: HashMap::default(),
-
-            string_rxs: vec![],
-            opt_string_rxs: vec![],
-            bool_rxs: vec![],
+        let child: View<Node> = self.slots.remove(index);
+        if cfg!(target_arch = "wasm32") {
+            let node: &Node = self.element.unchecked_ref();
+            node.remove_child(&child.dom_ref()).unwrap_throw();
         }
+        Some(child)
     }
-    #[cfg(target_arch = "wasm32")]
-    fn element_ns(tag: &str, ns: &str) -> Self {
-        let el = utils::document()
-            .create_element_ns(Some(ns), tag)
-            .expect(&format!("cannot create element_ns '{}' '{}'", tag, ns))
-            .unchecked_into();
-        View::wrapping(el)
-    }
-}
-
 
-impl<T: JsCast + AsRef<EventTarget>> EventTargetView for View<T> {
-    fn on(mut self, ev_name: &str, tx: Transmitter<Event>) -> View<T> {
+    pub fn remove_all_children(&mut self) -> Vec<View<Node>> {
+        let children: Vec<View<Node>> = std::mem::replace(&mut self.slots, vec![]);
         if cfg!(target_arch = "wasm32") {
-            let target: &EventTarget = self.as_ref();
-            let cb = utils::add_event(ev_name, target, tx);
-            self.callbacks.insert(ev_name.to_string(), cb);
-            self
-        } else {
-            self
+            let node: &Node = self.element.unchecked_ref();
+            for child in children.iter() {
+                node.remove_child(&child.dom_ref()).unwrap_throw();
+            }
         }
+        children
     }
 
-    fn window_on(mut self, ev_name: &str, tx: Transmitter<Event>) -> View<T> {
-        if cfg!(target_arch = "wasm32") {
-            let window = utils::window();
-            let target: &EventTarget = window.as_ref();
-            let cb = utils::add_event(ev_name, &target, tx);
-            self.window_callbacks.insert(ev_name.to_string(), cb);
-            self
-        } else {
-            self
+    /// Removes a child View node from a certain index and replaces it with another. Returns
+    /// the child view that was replaced.
+    /// If the index given is greater than the number of children the result is `None`.
+    pub fn replace_child_at(&mut self, index: usize, new_child: View<Node>) -> Option<View<Node>> {
```

Comment:
Noting that this function does not (currently) return the child view that was replaced (`old_child`) but rather returns the child view with which `old_child` was replaced (`new_child`). Or is the assertion that the swap of `ViewInternals` from `old_child` to `new_child` makes the semantics of "returns the child view that was replaced" accurate?

Reply:
You've hit on an important detail - a `View<T>` is a very thin wrapper over an internally mutable `VIewInternals`. Indeed the `swap` makes the semantics accurate, even though the memory location of the outer `View<Node>` wrapper coming out of the function will be the same as the one passed into it.

Reply:
Thanks for confirming. I've been enjoying poking around in the guts of what you've written, thanks for explaining and answering questions!